### PR TITLE
Optimize and clean up ProjectParser

### PR DESCRIPTION
**Pull request by jeffkl**
* Use HashSet<string> instead of string[]
* Remove small single use methods
* Add ReservedItemNames

### CODE REVIEW

**Comment by jeffkl** [comment]
@dotnet-bot test Windows_NT Build for CoreCLR please

**Comment by jeffkl** [comment]
@dotnet-bot test Windows_NT Build for Full please

File: src/Build/Evaluation/ProjectParser.cs [hunk 1]
```diff
@@ -262,40 +236,32 @@ private void ParseProjectRootElementChildren(XmlElementWithLocation element)
...
                 propertyGroup.AppendParentedChildNoChecks(property);
             }
 
             return propertyGroup;
         }
 
-        /// <summary>
-        /// Parse a ProjectPropertyElement from the element
-        /// </summary>
-        private ProjectPropertyElement ParseProjectPropertyElement(XmlElementWithLocation element, ProjectPropertyGroupElement parent)
```
Comment (cdmihai) [inline src/Build/Evaluation/ProjectParser.cs]: I wouldn't inline the parse methods for each element type, as they offer some conceptual modularization.

Comment (cdmihai) [reply src/Build/Evaluation/ProjectParser.cs]: And the compiler will hopefully inline it. Maybe you could add `[MethodImpl(MethodImplOptions.AggressiveInlining)]` to the methods instead?

File: src/Build/Evaluation/ProjectParser.cs [hunk 2]
```diff
@@ -562,46 +528,26 @@ private UsingTaskParameterGroupElement ParseUsingTaskParameterGroupElement(XmlEl
             return parameterGroup;
         }
 
-        /// <summary>
-        /// Parse a UsingTaskBodyElement from the element
-        /// </summary>
-        private ProjectUsingTaskBodyElement ParseUsingTaskBodyElement(XmlElementWithLocation element, ProjectUsingTaskElement parent)
```
Comment (cdmihai) [inline src/Build/Evaluation/ProjectParser.cs]: Same with these, I like it that each parse method is responsible for its own validation checks and interpretation.

File: src/Shared/XMakeAttributes.cs [hunk 3]
```diff
@@ -123,11 +123,7 @@ string attribute
         /// <returns>true, if the given attribute is reserved and badly cased</returns>
         internal static bool IsBadlyCasedSpecialTaskAttribute(string attribute)
         {
-            return (!IsSpecialTaskAttribute(attribute) &&
-                ((String.Compare(attribute, condition, StringComparison.OrdinalIgnoreCase) == 0) ||
-                (String.Compare(attribute, continueOnError, StringComparison.OrdinalIgnoreCase) == 0) ||
-                (String.Compare(attribute, msbuildRuntime, StringComparison.OrdinalIgnoreCase) == 0) ||
-                (String.Compare(attribute, msbuildArchitecture, StringComparison.OrdinalIgnoreCase) == 0)));
+            return !IsSpecialTaskAttribute(attribute) && KnownSpecialTaskAttributesIgnoreCase.Contains(attribute);
```
Comment (cdmihai) [inline src/Shared/XMakeAttributes.cs]: >at [](start = 43, length = 2)

Could you prepend the fields with "_" to make them consistent with most of msbuild?

File: src/Shared/XMakeAttributes.cs [hunk 4]
```diff
@@ -94,26 +91,29 @@ internal struct MSBuildArchitectureValues
         /////////////////////////////////////////////////////////////////////////////////////////////
         internal const string defaultXmlNamespace = "http://schemas.microsoft.com/developer/msbuild/2003";
 
+        private static readonly HashSet<string> KnownSpecialTaskAttributes = new HashSet<string> { condition, continueOnError, msbuildRuntime, msbuildArchitecture, xmlns };
+
+        private static readonly HashSet<string> KnownSpecialTaskAttributesIgnoreCase = new HashSet<string>(StringComparer.OrdinalIgnoreCase) { condition, continueOnError, msbuildRuntime, msbuildArchitecture, xmlns };
```
Comment (cdmihai) [inline src/Shared/XMakeAttributes.cs]: Could you prepend the fields with "_" to make them consistent with most of msbuild?

**Comment by AndyGerlicher** [comment]
The perf suite has gotten a little out of hand... But I think it looks good? @cdmihai 

### DesignTimeBuild Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 43.9811 -> 44.0742 (0.212%)
DotnetWebProject | ::ok_hand: | no | 267.4616 -> 265.0529 (-0.901%)
DotnetMvcProject | ::ok_hand: | no | 273.9715 -> 271.0847 (-1.054%)
Picasso | :red_circle: | yes | 1690.8512 -> 1702.5797 (0.694%)
SmallP2POldCsproj | :red_circle: | yes | 70.1541 -> 71.0864 (1.329%)
SmallP2PNewCsproj | :ok_hand: | no | 735.7655 -> 736.129 (0.049%)
OrchardCore | :white_check_mark: | yes | 55141.5154 -> 54998.3841 (-0.26%)
WebLargeCore | :white_check_mark: | yes | 61063.5414 -> 59971.9497 (-1.788%)

### DesignTimeBuild Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 2304296 -> 2233551 (-3.07%)
DotnetWebProject | ::ok_hand: | no | 14479086 -> 14427694 (-0.355%)
DotnetMvcProject | ::ok_hand: | no | 14612992 -> 14550548 (-0.427%)
Picasso | ::ok_hand: | no | 93185818 -> 93087776 (-0.105%)
SmallP2POldCsproj | ::ok_hand: | no | 8588824 -> 8587083 (-0.02%)
SmallP2PNewCsproj | :red_circle: | yes | 44012857 -> 44099993 (0.198%)
OrchardCore | :ok_hand: | no | 1352659590 -> 1358884477 (0.46%)
WebLargeCore | :red_circle: | yes | 1394798660 -> 1413550891 (1.344%)

### ParallelEvaluationIsolatedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :ok_hand: | no | 42.6649 -> 42.6826 (0.041%)
DotnetWebProject | :red_circle: | yes | 58.5196 -> 59.558 (1.774%)
DotnetMvcProject | :red_circle: | yes | 63.9803 -> 65.3003 (2.063%)
Picasso | ::ok_hand: | no | 165.2992 -> 165.246 (-0.032%)
SmallP2POldCsproj | :white_check_mark: | yes | 36.082 -> 35.9208 (-0.447%)
SmallP2PNewCsproj | ::ok_hand: | no | 105.3974 -> 104.4669 (-0.883%)
Generated_100_100_v150 | ::ok_hand: | no | 670.1582 -> 667.4147 (-0.409%)
OrchardCore | :white_check_mark: | yes | 1165.2913 -> 1154.8802 (-0.893%)
Roslyn | ::ok_hand: | no | 1494.9312 -> 1492.9826 (-0.13%)
WebLargeCore | :ok_hand: | no | 1351.1999 -> 1359.5249 (0.616%)

### ParallelEvaluationIsolatedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 2963734 -> 2964162 (0.014%)
DotnetWebProject | :white_check_mark: | yes | 4907922 -> 4905910 (-0.041%)
DotnetMvcProject | :white_check_mark: | yes | 5615827 -> 5614120 (-0.03%)
Picasso | ::ok_hand: | no | 12692103 -> 12219018 (-3.727%)
SmallP2POldCsproj | :ok_hand: | no | 4525973 -> 4537912 (0.264%)
SmallP2PNewCsproj | :ok_hand: | no | 8900675 -> 8914261 (0.153%)
Generated_100_100_v150 | ::ok_hand: | no | 34876689 -> 34492706 (-1.101%)
OrchardCore | ::ok_hand: | no | 62351791 -> 61377838 (-1.562%)
Roslyn | :ok_hand: | no | 88884967 -> 89121085 (0.266%)
WebLargeCore | :white_check_mark: | yes | 67587171 -> 63338182 (-6.287%)

### ParallelEvaluationSharedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 42.5155 -> 42.9588 (1.043%)
DotnetWebProject | :white_check_mark: | yes | 59.1901 -> 59.0705 (-0.202%)
DotnetMvcProject | :red_circle: | yes | 64.8074 -> 65.2705 (0.715%)
Picasso | ::ok_hand: | no | 167.9819 -> 166.2142 (-1.052%)
SmallP2POldCsproj | :white_check_mark: | yes | 36.0039 -> 35.9035 (-0.279%)
SmallP2PNewCsproj | ::ok_hand: | no | 102.127 -> 102.045 (-0.08%)
Generated_100_100_v150 | :ok_hand: | no | 666.4449 -> 667.6923 (0.187%)
OrchardCore | :ok_hand: | no | 1135.789 -> 1144.4587 (0.763%)
Roslyn | :white_check_mark: | yes | 1472.4934 -> 1464.9234 (-0.514%)
WebLargeCore | ::ok_hand: | no | 1336.3351 -> 1326.7946 (-0.714%)

### ParallelEvaluationSharedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 2962816 -> 2962564 (-0.009%)
DotnetWebProject | :ok_hand: | no | 4904339 -> 4904610 (0.006%)
DotnetMvcProject | :red_circle: | yes | 5612663 -> 5614985 (0.041%)
Picasso | :red_circle: | yes | 11345307 -> 11829189 (4.265%)
SmallP2POldCsproj | ::ok_hand: | no | 4554915 -> 4544653 (-0.225%)
SmallP2PNewCsproj | ::ok_hand: | no | 9262165 -> 9212131 (-0.54%)
Generated_100_100_v150 | :ok_hand: | no | 34569911 -> 35190980 (1.797%)
OrchardCore | ::ok_hand: | no | 61505379 -> 60379194 (-1.831%)
Roslyn | ::ok_hand: | no | 87990206 -> 87467677 (-0.594%)
WebLargeCore | :red_circle: | yes | 63067111 -> 66563404 (5.544%)

### SerialEvaluationIsolatedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 36.4567 -> 36.4549 (-0.005%)
DotnetWebProject | :red_circle: | yes | 58.6354 -> 59.0879 (0.772%)
DotnetMvcProject | :white_check_mark: | yes | 65.0064 -> 64.8198 (-0.287%)
Picasso | :red_circle: | yes | 340.9438 -> 345.1108 (1.222%)
SmallP2POldCsproj | :red_circle: | yes | 61.354 -> 62.0876 (1.196%)
SmallP2PNewCsproj | :red_circle: | yes | 245.3046 -> 247.1208 (0.74%)
Generated_100_100_v150 | :red_circle: | yes | 1686.9496 -> 1714.8297 (1.653%)
OrchardCore | :ok_hand: | no | 3476.3532 -> 3477.1814 (0.024%)
Roslyn | :red_circle: | yes | 4106.5792 -> 4131.9059 (0.617%)
WebLargeCore | :white_check_mark: | yes | 3975.876 -> 3877.0531 (-2.486%)

### SerialEvaluationIsolatedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | ::ok_hand: | no | 5565794 -> 5564961 (-0.015%)
DotnetWebProject | :ok_hand: | no | 3941554 -> 3946203 (0.118%)
DotnetMvcProject | :ok_hand: | no | 4585756 -> 4602833 (0.372%)
Picasso | ::ok_hand: | no | 14294536 -> 14209654 (-0.594%)
SmallP2POldCsproj | ::ok_hand: | no | 3266425 -> 3262317 (-0.126%)
SmallP2PNewCsproj | :red_circle: | yes | 8067545 -> 8455271 (4.806%)
Generated_100_100_v150 | :white_check_mark: | yes | 34755631 -> 34000040 (-2.174%)
OrchardCore | ::ok_hand: | no | 59301856 -> 58963008 (-0.571%)
Roslyn | ::ok_hand: | no | 83576461 -> 83386899 (-0.227%)
WebLargeCore | ::ok_hand: | no | 58965817 -> 57951521 (-1.72%)

### SerialEvaluationSharedContextSecondRun Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 34.3543 -> 34.5821 (0.663%)
DotnetWebProject | :white_check_mark: | yes | 62.4753 -> 62.2835 (-0.307%)
DotnetMvcProject | :red_circle: | yes | 65.1219 -> 65.1857 (0.098%)
Picasso | :white_check_mark: | yes | 345.8084 -> 344.6029 (-0.349%)
SmallP2POldCsproj | :red_circle: | yes | 63.8483 -> 63.9116 (0.099%)
SmallP2PNewCsproj | :red_circle: | yes | 233.1179 -> 234.8466 (0.742%)
Generated_100_100_v150 | :red_circle: | yes | 1657.5376 -> 1676.1736 (1.124%)
OrchardCore | :red_circle: | yes | 3377.7389 -> 3394.5033 (0.496%)
Roslyn | :white_check_mark: | yes | 4020.9634 -> 3958.0597 (-1.564%)
WebLargeCore | :ok_hand: | no | 3680.389 -> 3688.2271 (0.213%)

### SerialEvaluationSharedContextSecondRun Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 5695588 -> 5692627 (-0.052%)
DotnetWebProject | :white_check_mark: | yes | 4077047 -> 4076393 (-0.016%)
DotnetMvcProject | :white_check_mark: | yes | 4823437 -> 4816530 (-0.143%)
Picasso | :red_circle: | yes | 14469886 -> 14470869 (0.007%)
SmallP2POldCsproj | :white_check_mark: | yes | 3346669 -> 3346666 (0%)
SmallP2PNewCsproj | :ok_hand: | no | 7276606 -> 7299820 (0.319%)
Generated_100_100_v150 | :white_check_mark: | yes | 33826474 -> 33628953 (-0.584%)
OrchardCore | :white_check_mark: | yes | 59404463 -> 58720677 (-1.151%)
Roslyn | ::ok_hand: | no | 85719953 -> 85677428 (-0.05%)
WebLargeCore | ::ok_hand: | no | 60149712 -> 60104225 (-0.076%)

### SerialEvaluationSharedContext Time (ms) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :red_circle: | yes | 42.9899 -> 43.1505 (0.374%)
DotnetWebProject | :red_circle: | yes | 58.2905 -> 58.426 (0.232%)
DotnetMvcProject | :red_circle: | yes | 63.8386 -> 63.9649 (0.198%)
Picasso | :white_check_mark: | yes | 346.5024 -> 344.3411 (-0.624%)
SmallP2POldCsproj | :red_circle: | yes | 60.3336 -> 60.5114 (0.295%)
SmallP2PNewCsproj | :red_circle: | yes | 231.0796 -> 233.1071 (0.877%)
Generated_100_100_v150 | :ok_hand: | no | 1684.1134 -> 1684.457 (0.02%)
OrchardCore | :red_circle: | yes | 3410.4118 -> 3426.3788 (0.468%)
Roslyn | :white_check_mark: | yes | 4021.1474 -> 3985.4403 (-0.888%)
WebLargeCore | :red_circle: | yes | 3690.2916 -> 3730.7996 (1.098%)

### SerialEvaluationSharedContext Memory (bytes) (src\msbuild.A vs src\msbuild.B)
Test | Overall | Significant δ  | Value
:--- | :--- | :--- | :---
DotnetConsoleProject | :white_check_mark: | yes | 2962710 -> 2962553 (-0.005%)
DotnetWebProject | :white_check_mark: | yes | 4904520 -> 4904193 (-0.007%)
DotnetMvcProject | :ok_hand: | no | 5613966 -> 5614914 (0.017%)
Picasso | :white_check_mark: | yes | 10183858 -> 10177423 (-0.063%)
SmallP2POldCsproj | :red_circle: | yes | 4281730 -> 4283514 (0.042%)
SmallP2PNewCsproj | ::ok_hand: | no | 8035435 -> 8021506 (-0.173%)
Generated_100_100_v150 | :white_check_mark: | yes | 34034176 -> 33807313 (-0.667%)
OrchardCore | :ok_hand: | no | 58617686 -> 58723779 (0.181%)
Roslyn | ::ok_hand: | no | 83171571 -> 83087003 (-0.102%)
WebLargeCore | :ok_hand: | no | 59539061 -> 59960281 (0.707%)

File: src/Build/Evaluation/ProjectParser.cs (hunk 1 above)
Comment (jeffkl) [reply src/Build/Evaluation/ProjectParser.cs]: What I found when I was going through it was a lot of bouncing back and forth.  I found it hard to clean up with so much of the logic spread all over the place.  Some of the methods are part of parsing the children of an element so I thought it made sense to reduce complexity.

File: src/Shared/XMakeAttributes.cs (hunk 4 above)
Comment (jeffkl) [reply src/Shared/XMakeAttributes.cs]: `static readonly` members should look like fields and properties with no leading underscore and proper case.

**Pull request merged**